querybook/server/clients/github_client.py

```diff
@@ -0,0 +1,130 @@
+
+LOG = get_logger(__name__)
+
+
+class GitHubClient:
+    def __init__(self, github_link: GitHubLink, datadoc: DataDoc):
+        """
+        Initialize the GitHub client with an access token from the session.
+        Raises an exception if the token is not found.
+        """
+        access_token = self._get_access_token()
+        auth = Auth.Token(access_token)
+        self.client = Github(auth=auth)
+        self.user = self.client.get_user()
+        self.github_link = github_link
+        self.repo = self._get_repository()
+        self.branch = "main"
+        self.file_path = self._build_file_path(datadoc)
+
+    def _get_access_token(self) -> str:
+        access_token = flask_session.get("github_access_token")
+        if not access_token:
+            LOG.error("GitHub OAuth token not found in session")
+            raise Exception("GitHub OAuth token not found in session")
+        return access_token
+
+    def _get_repository(self):
+        repo_url = QuerybookSettings.GITHUB_REPO_URL
+        if not repo_url:
+            LOG.error("GITHUB_REPO_URL is not configured")
+            raise Exception("GITHUB_REPO_URL is not configured")
+        repo_full_name = self._extract_repo_full_name(repo_url)
+        return self.client.get_repo(repo_full_name)
+
+    @staticmethod
+    def _extract_repo_full_name(repo_url: str) -> str:
+        # Assumes repo_url is in the format 'https://github.com/owner/repo'
+        parts = repo_url.rstrip("/").split("/")
+        if len(parts) >= 2:
+            return f"{parts[-2]}/{parts[-1]}"
+        else:
+            raise ValueError("Invalid GITHUB_REPO_URL configuration")
+
+    def _build_file_path(self, datadoc: DataDoc) -> str:
+        directory = self.github_link.directory
+        file_name = f"datadoc_{datadoc.id}.md"
+        return f"{directory}/{file_name}"
+
+    def commit_datadoc(self, datadoc: DataDoc, commit_message: Optional[str] = None):
+        """
+        Commit a DataDoc to the repository.
+        Args:
+            datadoc (DataDoc): The DataDoc object to commit.
+            commit_message (Optional[str]): Commit message. Defaults to a standard message.
+        Raises:
+            Exception: If committing the DataDoc fails.
+        """
+        content = serialize_datadoc_to_markdown(datadoc)
+        if not commit_message:
+            commit_message = (
+                f"Update DataDoc {datadoc.id}: {datadoc.title or 'Untitled'}"
+            )
+
+        try:
+            contents = self.repo.get_contents(self.file_path, ref=self.branch)
+            # Update file
+            self.repo.update_file(
+                path=contents.path,
+                message=commit_message,
+                content=content,
+                sha=contents.sha,
+                branch=self.branch,
+            )
+            LOG.info(f"Updated file {self.file_path} in repository.")
+        except GithubException as e:
+            if e.status == 404:
```

Comment:
if this the recommended way to check the file existence?

Reply:
I think so since we can handle creating and updating file in 1 api call. We check for the existence of a file in a GitHub repository, and create new one if None exist. For any other error code we treat as regular exception.

Reply:
then you'll need to also handle the exception of `create_file`